### PR TITLE
Add tests for IStructuralEquatable/Comparable throwing NullReferenceException

### LINKED ISSUE #13410: Tuple.IStructuralComparable/Equatable throws NRE
The following tests demonstrates this.
```
[Fact]
public void IStructuralEquatable_Equals_NullEqualityComparer_ThrowsNullReferenceException()
{
    IStructuralEquatable structuralEquatable = Tuple.Create(1);
    Assert.Throws<NullReferenceException>(() => structuralEquatable.Equals(Tuple.Create(1), null));
}

[Fact]
public void IStructuralEquatable_GetHashCode_NullEqualityComparer_ThrowsNullReferenceException()
{
    IStructuralEquatable structuralEquatable = Tuple.Create(1);
    Assert.Throws<NullReferenceException>(() => structuralEquatable.GetHashCode(null));
}

[Fact]
public void IStructuralComparable_CompareTo_NullEqualityComparer_ThrowsNullReferenceException()
{
    IStructuralComparable structuralComparable = Tuple.Create(1);
    Assert.Throws<NullReferenceException>(() => structuralComparable.CompareTo(Tuple.Create(1), null));
}
```

I can send in a PR fixing this to throw ArgumentNullException, but there might be breaking change so I'll let you guys approve it before I go wasting my time ;)

### LINKED ISSUE #13429: ValueTuple.IStructuralComparable/Equatable throws NRE instead of ArgumentNullException
The following tests demonstrates this.
```
[Fact]
public void IStructuralEquatable_Equals_NullEqualityComparer_ThrowsNullReferenceException()
{
    IStructuralEquatable structuralEquatable = ValueTuple.Create(1);
    Assert.Throws<NullReferenceException>(() => structuralEquatable.Equals(Tuple.Create(1), null));
}

[Fact]
public void IStructuralEquatable_GetHashCode_NullEqualityComparer_ThrowsNullReferenceException()
{
    IStructuralEquatable structuralEquatable = ValueTuple.Create(1);
    Assert.Throws<NullReferenceException>(() => structuralEquatable.GetHashCode(null));
}

[Fact]
public void IStructuralComparable_CompareTo_NullEqualityComparer_ThrowsNullReferenceException()
{
    IStructuralComparable structuralComparable = ValueTuple.Create(1);
    Assert.Throws<NullReferenceException>(() => structuralComparable.CompareTo(Tuple.Create(1), null));
}
```

Spun off from #13410

/cc @karelz (who can help triage this :) )

### PR DESCRIPTION
**Pull request by hughbe**
Closes #13410
Closes #13429

/cc @stephentoub @jcouv

### CODE REVIEW

**Comment by jcouv** [comment]
@stephentoub My intuition is that changing the thrown exception from `System.Tuple` APIs is a breaking change. Is that correct?
If we do decide to make this change, we should flag or track this to make the same change to various mscorlibs that are not public (.Net desktop).

FYI @KevinRansom This PR proposes to change the exception thrown by both System.Tuple and ValueTuple.

**Comment by stephentoub** [comment]
> My intuition is that changing the thrown exception from System.Tuple APIs is a breaking change. Is that correct?

Correct

> This PR proposes to change the exception thrown by both System.Tuple and ValueTuple.

Has the PR been updated?  It looks like it's only adding tests at this point, not changing any exception-throwing logic in implementation.

File: src/System.Runtime/tests/System/TupleTests.cs [hunk 1]
```diff
@@ -210,13 +210,39 @@ public void Equals_GetHashCode(TupleTestDriver<T1, T2, T3, T4, T5, T6, T7, T8, T
...
             public void CompareTo(TupleTestDriver<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10> other, int expectedResult, int expectedStructuralResult)
             {
                 Assert.Equal(expectedResult, ((IComparable)Tuple).CompareTo(other.Tuple));
                 Assert.Equal(expectedStructuralResult, ((IStructuralComparable)Tuple).CompareTo(other.Tuple, TestComparer.Instance));
                 Assert.Equal(1, ((IComparable)Tuple).CompareTo(null));
+
+                IStructuralComparable_NullComparer_ThrowsNullReferenceException();  
+            }
+
+            public void IStructuralComparable_NullComparer_ThrowsNullReferenceException()
```
Comment (jcouv) [inline src/System.Runtime/tests/System/TupleTests.cs]: Consider adding a `[Fact]` on this so it runs as its own test (no need to invoke it from method above).

**Comment by jcouv** [comment]
My bad. I got confused by the comment saying this fixes linked issues. Thanks for the correction.

The tests LGTM, although I'd suggest tagging the new methods as their own [Fact]. Thanks

**Comment by hughbe** [comment]
@jcouv my fault too - I just updated the description to say "closes" instead of "fixes"

File: src/System.Runtime/tests/System/TupleTests.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Runtime/tests/System/TupleTests.cs]: Currently this is not possible, as no classes inherit from TupleTestDriver, so the `[Fact]` test will never run.

If you see the meat of the `Tuple` tests below, we have to call each test manually

Comment (hughbe) [inline src/System.Runtime/tests/System/TupleTests.cs]: Obvi, this should be cleaned up/fixed up, but for another PR

**Comment by jcouv** [comment]
@hughbe Thanks!

**Pull request merged**